readme.txt
=== GD Plugin Core ===
Contributors: gdragon
Donate link: http://www.dev4press.com/donate/
Version: 1.1.1
Tags: widget, plugin, multi, options, base, class, core, extend
Requires at least: 2.5
Tested up to: 2.7.1
Stable tag: trunk

This is not really a plugin. This is a base class implementation that can be used to create plugins.

== Description ==

Basic idea was to shorten the plugin development time. When you start working on a new plugin, you need to create a lot of basic code for initialization of options page, widget  initialization and this is even more complicated if you want a plugin to support multi instance widgets. And this you need to repeat for every plugin.

And, repeating this every time gets very tiresome and mistakes are easily made. So, you might want to try using this base class, extending it and setting only the things you need.

= Class Features =

The GD Plugin Core class implements following features:

* Options subpage on a Settings page
* Options subpage on a Plugins page
* Options subpage on a Manage page
* Setting user level for added subpages
* Multi instance widget
* Loading of translations if available
* Adding files (css, js...) into header
* Event log with log file support
* WordPress filters for header and footer
* Unlimited number of Shortcodes
* Additional Helper class with useful functions

== Installation ==

* Upload `gd-plugin-core` folder to the `/wp-content/plugins/` directory
* Activate the plugin through the 'Plugins' menu in WordPress

== Frequently Asked Questions ==

= How to use it?
There is an user guide included.

== Website ==

* Homepage: http://www.dev4press.com/
* Plugin: http://www.dev4press.com/plugins/gd-plugin-core/
* Old Blog: http://wp.gdragon.info/

== Support ==

* Ohloh: https://www.ohloh.net/p/gd-plugin-core
* Forum: http://forum.gdragon.info/viewforum.php?f=17
* User Guide: http://wp.gdragon.info/plugins/gd-plugin-core/userguide/
* Email: wordpress@gdragon.info

== WordPress Extend ==

* Plugin: http://wordpress.org/extend/plugins/gd-plugin-core/

== Screenshots ==

1. Plugin Creation Wizard


changelog.txt
= GD Plugin Core =

== Change Log ==

=== Version: 1.1.1 / march 21 2009 ===
  * edit: updated readme file

=== Version: 1.1.0 / january 12 2009 ===
  * new: gdragon install db and debug classes
  * new: resize image function
  * new: phpdoc documentation

=== Version: 1.0.3 / december 11 2008 ===
  * new: wizard translation support
  * edit: wordpress 2.7 visual changes
  * fix: some language references

=== Version: 1.0.2 / july 31 2008 ===
  * Initial Public Release


gd-plugin-form.php
<div class="wrap"><h2>GD Plugin Core: Wizard</h2>

<?php 

function render_user_level_combo() {
?>
<option value="10">10: <?php _e("Admin Only", "gd-plugin-core"); ?></option>
<option value="09">09: <?php _e("Administrator", "gd-plugin-core"); ?></option>
<option value="08">08: <?php _e("Editor", "gd-plugin-core"); ?></option>
<option value="07">07: <?php _e("Editor", "gd-plugin-core"); ?></option>
<option value="06">06: <?php _e("Editor", "gd-plugin-core"); ?></option>
<option value="05">05: <?php _e("Editor", "gd-plugin-core"); ?></option>
<option value="04">04: <?php _e("Editor", "gd-plugin-core"); ?></option>
<option value="03">03: <?php _e("Author", "gd-plugin-core"); ?></option>
<option value="02">02: <?php _e("Contributor", "gd-plugin-core"); ?></option>
<option value="01">01: <?php _e("Contributor", "gd-plugin-core"); ?></option>
<?php
}

if (!$writeable) {
    echo '<div class="updated fade" id="message" style="background-color: rgb(255, 251, 204); width: 980px; margin-top: 10px"><p>';
    _e("Folder <strong>plugins-wizard</strong> needs to be writeable for wizard to automaticlly create all plugin folders and files.", "gd-plugin-core");
    echo '<br />';
    _e("You must fix this before you can use this plugin. Only then <strong>Create Plugin</strong> button will be visible.", "gd-plugin-core");
    echo '</p></div>'; 
}

if ($status != '') {
    echo '<div class="updated fade" id="message" style="background-color: rgb(255, 251, 204); width: 980px; margin-top: 10px"><p>'.$status.'</p></div>';
}

?>

<form method="post">
<input type="hidden" id="gdcc-action" name="gdcc-action" value="save">
<table class="form-table">
<tbody>
<tr valign="top">
<th scope="row"><?php _e("Plugin Details", "gd-plugin-core"); ?></th>
<td>
<table>
    <tr>
        <td width="100" style="padding: 0; border: 0;"><?php _e("Name", "gd-plugin-core"); ?>:</td>
        <td style="padding: 0; border: 0;"><input name="gdcc[name]" type="text" id="gdcc-name" value="GD Example Plugin" size="32" /></td>
    </tr>
    <tr>
        <td width="100" style="padding: 0; border: 0;"><?php _e("Description", "gd-plugin-core"); ?>:</td>
        <td style="padding: 0; border: 0;"><input name="gdcc[description]" type="text" id="gdcc-description" value="This is an plugin example." size="32" /></td>
    </tr>
    <tr>
        <td width="100" style="padding: 0; border: 0;"><?php _e("Plugin URL", "gd-plugin-core"); ?>:</td>
        <td style="padding: 0; border: 0;"><input name="gdcc[url]" type="text" id="gdcc-url" value="<?php bloginfo('home'); ?>/" size="32" /></td>
    </tr>
    <tr>
        <td width="100" style="padding: 0; border: 0;"><?php _e("Author", "gd-plugin-core"); ?>:</td>
        <td style="padding: 0; border: 0;"><input name="gdcc[author]" type="text" id="gdcc-author" value="Your Name" size="32" /></td>
    </tr>
    <tr>
        <td width="100" style="padding: 0; border: 0;"><?php _e("Author URL", "gd-plugin-core"); ?>:</td>
        <td style="padding: 0; border: 0;"><input name="gdcc[author_url]" type="text" id="gdcc-author_url" value="<?php bloginfo('home'); ?>/" size="32" /></td>
    </tr>
</table>
</td>
</tr>
<tr valign="top">
<th scope="row"><?php _e("Features", "gd-plugin-core"); ?></th>
<td>
<label for="gdcc[widget]">
    <input name="gdcc[widget]" type="checkbox" id="gdcc-widget" value="1"  checked="checked" /> <?php _e("Multi instance Widget", "gd-plugin-core"); ?>
</label><br />
<label for="gdcc[shortcode]">
    <input name="gdcc[shortcode]" type="checkbox" id="gdcc-shortcode" value="1"  checked="checked" /> <?php _e("Shortcode support", "gd-plugin-core"); ?>
</label><br />
<label for="gdcc[i18n]">
    <input name="gdcc[i18n]" type="checkbox" id="gdcc-i18n" value="1"  checked="checked" /> <?php _e("Multi language support", "gd-plugin-core"); ?>
</label>
</td>
</tr>
<tr valign="top">
<th scope="row"><?php _e("Administration Panels", "gd-plugin-core"); ?></th>
<td>
<table>
    <tr>
        <td width="110" style="padding: 0; border: 0;">
        <label for="gdcc[admin_options]">
            <input name="gdcc[admin_options]" type="checkbox" id="gdcc-admin_options" value="1"  checked="checked" /> <?php _e("Settings Panel", "gd-plugin-core"); ?>
        </label>
        <td width="70" style="padding: 0; border: 0;"><?php _e("User Level", "gd-plugin-core"); ?>:</td>
        </td>
        <td style="padding: 0; border: 0;"><select name="gdcc[admin_options_level]" id="gdcc-admin_options_level"><?php render_user_level_combo(); ?></select></td>
    </tr>
    <tr>
        <td width="110" style="padding: 0; border: 0;">
        <label for="gdcc[admin_plugins]">
            <input name="gdcc[admin_plugins]" type="checkbox" id="gdcc-admin_plugins" value="1" /> <?php _e("Plugins Panel", "gd-plugin-core"); ?>
        </label>
        <td width="70" style="padding: 0; border: 0;"><?php _e("User Level", "gd-plugin-core"); ?>:</td>
        </td>
        <td style="padding: 0; border: 0;"><select name="gdcc[admin_plugins_level]" id="gdcc-admin_plugins_level"><?php render_user_level_combo(); ?></select></td>
    </tr>
    <tr>
        <td width="110" style="padding: 0; border: 0;">
        <label for="gdcc[admin_options]">
            <input name="gdcc[admin_manage]" type="checkbox" id="gdcc-admin_manage" value="1" /> <?php _e("Manage Panel", "gd-plugin-core"); ?>
        </label>
        <td width="70" style="padding: 0; border: 0;"><?php _e("User Level", "gd-plugin-core"); ?>:</td>
        </td>
        <td style="padding: 0; border: 0;"><select name="gdcc[admin_manage_level]" id="gdcc-admin_manage_level"><?php render_user_level_combo(); ?></select></td>
    </tr>
</table>
</td>
</tr>
<tr valign="top">
<th scope="row"><?php _e("Administration Pages", "gd-plugin-core"); ?></th>
<td>
<label for="gdcc[admin_header]">
    <input name="gdcc[admin_header]" type="checkbox" id="gdcc-admin_header" value="1"  checked="checked" /> <?php _e("Header", "gd-plugin-core"); ?>
</label><br />
<label for="gdcc[footer]">
    <input name="gdcc[admin_footer]" type="checkbox" id="gdcc-admin_footer" value="1"  checked="checked" /> <?php _e("Footer", "gd-plugin-core"); ?>
</label>
</td>
</tr>
<tr valign="top">
<th scope="row"><?php _e("Blog", "gd-plugin-core"); ?></th>
<td>
<label for="gdcc[wp_content]">
    <input name="gdcc[wp_content]" type="checkbox" id="gdcc-wp_content" value="1"  checked="checked" /> <?php _e("Wordpress Loop content manipulation", "gd-plugin-core"); ?>
</label><br />
<label for="gdcc[wp_header]">
    <input name="gdcc[wp_header]" type="checkbox" id="gdcc-wp_header" value="1"  checked="checked" /> <?php _e("Header", "gd-plugin-core"); ?>
</label><br />
<label for="gdcc[wp_footer]">
    <input name="gdcc[wp_footer]" type="checkbox" id="gdcc-wp_footer" value="1"  checked="checked" /> <?php _e("Footer", "gd-plugin-core"); ?>
</label>
</td>
</tr>
<tr valign="top">
<th scope="row"><?php _e("Plugin", "gd-plugin-core"); ?></th>
<td>
<table>
    <tr>
        <td width="100" style="padding: 0; border: 0;">Options name:</td>
        <td style="padding: 0; border: 0;"><input name="gdcc[plugin_options]" type="text" id="gdcc-plugin_options" value="" size="32" /></td>
    </tr>
</table>
<?php _e("Leave blank if you want base class to generate this name based on your plugin name. Do not use blank spaces.", "gd-plugin-core"); ?>
</td>
</tr>
<tr valign="top">
<th scope="row">Widget</th>
<td>
<table>
    <tr>
        <td width="100" style="padding: 0; border: 0;"><?php _e("Options name", "gd-plugin-core"); ?>:</td>
        <td style="padding: 0; border: 0;"><input name="gdcc[widget_options]" type="text" id="gdcc-widget_options" value="" size="32" /></td>
    </tr>
    <tr>
        <td width="100" style="padding: 0; border: 0;"><?php _e("Widget Base ID", "gd-plugin-core"); ?>:</td>
        <td style="padding: 0; border: 0;"><input name="gdcc[widget_baseid]" type="text" id="gdcc-widget_baseid" value="" size="32" /></td>
    </tr>
    <tr>
        <td width="100" style="padding: 0; border: 0;"><?php _e("Base Form Name", "gd-plugin-core"); ?>:</td>
        <td style="padding: 0; border: 0;"><input name="gdcc[widget_form]" type="text" id="gdcc-widget_form" value="" size="32" /></td>
    </tr>
</table>
<table>
<tr>
    <td width="100" style="padding: 0; border: 0;"><?php _e("Width", "gd-plugin-core"); ?>:</td>
    <td style="padding: 0; border: 0;"><input name="gdcc[widget_width]" type="text" id="gdcc-widget_width" value="400" size="5" /></td>
    <td width="18" style="padding: 0; border: 0;"></td>
    <td width="100" style="padding: 0; border: 0;"><?php _e("Height", "gd-plugin-core"); ?>:</td>
    <td style="padding: 0; border: 0;"><input name="gdcc[widget_height]" type="text" id="gdcc-widget_height" value="300" size="5" /></td>
</tr>
</table>
<?php _e("Leave blank if you want base class to generate name and id based on your plugin name. Do not use blank spaces.", "gd-plugin-core"); ?>
</td>
</tr>
</tbody>
</table>
<?php if ($writeable) : ?>
<p class="submit">
    <input type="submit" value="<?php _e("Create Plugin", "gd-plugin-core"); ?>" name="Submit"/>
</p>
<?Php endif; ?>
</form>
</div>

gd-plugin-xtra.php
<?php
/*
Name: GD Plugin Core: Xtras
Description: Help functions
*/

class GDCoreXtra {
    var $img_default = array(
        'limit_extension' => 'gif|jpg|png',
        'limit_width' => 0,
        'limit_height' => 0,
        'auto_resize' => false
    );

    /**
     * Renders into string items for select html tag.
     *
     * @param array $options options to render
     * @param string $selected selected option value
     * @return string rendered options for select tag
     */
    function get_select_options($options, $selected = "") {
        $output = "";
        for ($i = 0; $i < count($options); $i++) {
            $o = $options[$i];
            if (is_array($o)) {
                $value = $o["value"];
                $text = $o["text"];
            }
            else {
                $value = $i + 1;
                $text = $o;
            }
            if ($selected == $value) $current = ' selected="selected"';
            else $current = '';
            $output.= "<option value='".$value."'".$current.">".$text."</option>\r\n";
        }
        return $output;
    }
    
    /**
     * Renders and echoes string items for select html tag.
     *
     * @param array $options options to render
     * @param string $selected selected option value
     */
    function render_select_options($options, $selected = "") {
        echo get_render_select_options($options, $selected);
    }

    /**
     * Resizes the image.
     *
     * @param string $input file path to input file
     * @param <type> $output file path to output file
     * @param <type> $ext file extension
     * @param <type> $xs new width
     * @param <type> $ys new hieght
     */
    function resize_image($input, $output, $ext, $xs, $ys) {
        switch ($ext)
        {
            case "jpg":
                $src_image = imagecreatefromjpeg($input);
                break;
            case "png":
                $src_image = imagecreatefrompng($input);
                break;
            case "gif":
                $src_image = imagecreatefromgif($input);
                break;
        }

        $src_size = getimagesize($input);
        $xr = $src_size[0] / $xs;
        $yr = $src_size[1] / $ys;
        if ($xr < 1 && $yr < 1) {
            rename($input, $output);
        }
        else {
            if ($xr >= $yr) {
                $xn = $xs;
                $yn = floor($src_size[1] / $xr);
            }
            else {
                $yn = $ys;
                $xn = floor($src_size[0] / $yr);
            }
            $end_image = imagecreatetruecolor($xn, $yn);
            imagecopyresampled($end_image, $src_image, 0, 0, 0, 0, $xn, $yn, $src_size[0], $src_size[1]);
            imagedestroy($src_image);
            unlink($input);

            switch ($ext)
            {
                case "jpg":
                    imagejpeg($end_image, $output);
                    break;
                case "png":
                    imagepng($end_image, $output);
                    break;
                case "gif":
                    imagegif($end_image, $output);
                    break;
            }
    }
}

    function import_image($upl_file, $output_folder, $settings = array()) {
        $result = array();
        $image_path = "";
        if (is_uploaded_file($upl_file["tmp_name"])) {
            $exts = explode("|", $options["logo_ext"]);
            $upl_ext = strtolower(substr($upl_file["name"], -3, 3));
            if (in_array($upl_ext, $exts)) {

            }
        }
    }
}

?>

examples/gd-example-plugin-i/gd-example-plugin-i.php
<?php
/*
Plugin Name: GD Example Plugin I
Plugin URI: http://localhost:81/wpt/
Description: This is an plugin example.
Version: 1.0.0
Author: Your Name
Author URI: http://localhost:81/wpt/
*/

require_once (dirname(dirname(__FILE__)).'/gd-plugin-core/gd-plugin-core.php');
require_once (dirname(dirname(__FILE__)).'/gd-plugin-core/gd-plugin-xtra.php');

class GDExamplePluginI extends GDPluginCore
{
    var $shortcodes = array();
    var $default_options_widget = array(
        'title' => 'Example Title',
        'message' => 'This is an example plugin message.'
    );

    function GDExamplePluginI() {
        parent::GDPluginCore('GD Example Plugin I', __FILE__, 'widget_width=250&i18n=0&wp_head=0&wp_footer=0&wp_content=0');
    }

    function widget_control($widget_args = 1) {
        parent::widget_control($widget_args);
        include ("gd-example-plugin-i_widget.php");
    }

    function widget_parse_options($post_params) {
        $options = parent::widget_parse_options($post_params);
        
        $options["title"] = $post_params["title"];
        $options["message"] = $post_params["message"];

        return $options;
    }

    function widget_display($args, $widget_args = 1) {
        parent::widget_display($args, $widget_args);
        extract($args);
        $options = $this->options_widget;
        
        echo $before_widget.$before_title.$options['title'].$after_title;
        echo $options['message'].$after_widget;
    }
}

$pc_GDE = new GDExamplePluginI();

?>


gd-plugin-core.php
<?php
/*
Name: GD Plugin Core: Base
Description: Base class
*/

class GDPluginCore
{
    var $version = "1.1.1";
    
    var $log_file = "c:/log.txt";
    var $events;

    var $wp_version;
    var $plugin_arguments;
    var $plugin_name;
    var $plugin_id;
    var $plugin_url;
    var $plugin_path;
    var $file_path;
	
    var $extra_files_admin = array();
    var $extra_files_wp = array();
    var $shortcodes = array();
    
    var $global_options_widget;
    
    var $options_plugin;
    var $options_widget;
    var $options_widget_number;
    
    var $default_options_plugin;
    var $default_options_widget;
    
    function GDPluginCore($pl_name, $file, $arguments = "") {
        global $wp_version;
        $this->events = Array();

        // contruction //
        $this->plugin_name = $pl_name;
        $this->file_path = $file;
        $this->plugin_path = dirname($this->file_path);
        $this->plugin_id = basename(dirname($this->file_path));
        $this->set_params($arguments);
        
        $this->wp_version = substr(str_replace('.', '', $wp_version), 0, 2);
        if ($this->wp_version < 26) $this->plugin_url = get_option('home').'/'.PLUGINDIR.'/'.$this->plugin_id.'/';
        else $this->plugin_url = WP_PLUGIN_URL.'/'.$this->plugin_id.'/';

        // others //
        $this->global_options_widget = get_option($this->plugin_arguments["widgetoptions"]);

        // add actions & filters //
        if ($this->plugin_arguments["i18n"] == 1) add_action('init', array(&$this, 'lang_init'));
        if ($this->plugin_arguments["widget"] == 1) add_action('widgets_init', array(&$this, 'widget_init'));

        if ($this->plugin_arguments["admin_page_settings"] > 0) add_action('admin_menu', array(&$this, 'settings_menu')); 
        if ($this->plugin_arguments["admin_page_plugin"] > 0) add_action('admin_menu', array(&$this, 'plugin_menu')); 
        if ($this->plugin_arguments["admin_page_manage"] > 0) add_action('admin_menu', array(&$this, 'manage_menu')); 
        if ($this->plugin_arguments["admin_head"] == 1) add_action('admin_head', array(&$this, 'admin_head')); 
        if ($this->plugin_arguments["admin_footer"] == 1) add_action('admin_footer', array(&$this, 'admin_footer')); 
        if ($this->plugin_arguments["wp_head"] == 1) add_action('wp_head', array(&$this, 'wp_head'));
        if ($this->plugin_arguments["wp_head"] == 1) add_action('wp_footer', array(&$this, 'wp_footer'));
        if ($this->plugin_arguments["wp_content"] == 1) add_filter('the_content', array(&$this, 'wp_content'));
        
        if ($this->plugin_arguments["shortcodes"] == 1) {
            foreach ($this->shortcodes as $code) {
                $this->shortcode_action($code);
            }
        }
    }

    // helpers
    function set_params($arguments) {
        $p_name_1 = str_replace(" ", "-", strtolower($this->plugin_name));
        $p_name_2 = str_replace(" ", "", strtolower($this->plugin_name));
        
        $defaults = array(
            "widget" => 1,
            "i18n" => 1,
            "shortcodes" => 1,
            "admin_page_settings" => 10,
            "admin_page_plugin" => 0,
            "admin_page_manage" => 0,
            "admin_head" => 1,
            "admin_footer" => 1,
            "wp_head" => 1,
            "wp_footer" => 1,
            "wp_content" => 1,
            "plugin_options" => $p_name_1."_plugin",
            "widget_options" => $p_name_1."_widget",
            "widget_idbase" => $p_name_2."wimu",
            "widget_form" => $p_name_2."_wifo",
            "widget_width" => 400,
            "widget_height" => 300
        );

        $this->plugin_arguments = wp_parse_args($arguments, $defaults);
        extract($this->plugin_arguments, EXTR_SKIP);
    }
    
    function shortcode_action($scode) {
        if (is_array($scode)) {
            $sc_name = $scode["name"];
            $sc_method = $scode["function"];
        }
        else {
            $sc_name = $scode;
            $sc_method = "shortcode_".$scode;
        }
        add_shortcode($sc_name, array(&$this, $sc_method));
    }
    
    function create_file_tag($file) {
        if (is_array($file)) {
            $ffile = $file["file"];
            $fext = $file["ext"];
            if (isset($file["media"])) $fmedia = $file["media"];
            if (isset($file["type"])) $ftype = $file["type"];
            if (isset($frel["rel"])) $frel = $file["rel"];
        }
        else {
            $f = pathinfo($file);
            $ffile = $f["basename"];
            $fext = $f["extension"];
        }
        if (substr($ffile, 0, 1) == "/") $ffile = substr($ffile, 1);
        $ffile = $this->plugin_url.$ffile;
        $output = "";
        switch($fext) {
            case "js":
                $output = sprintf('<script type="%s" src="%s"></script>', 
                    isset($ftype) ? $ftype : 'text/javascript',
                    $ffile
                );
                break;
            default:
                $output = sprintf('<link rel="%s" href="%s" type="%s" media="%s" />', 
                    isset($frel) ? $frel : 'stylesheet',
                    $ffile,
                    isset($ftype) ? $ftype : 'text/css',
                    isset($fmedia) ? $fmedia : 'screen'
                );
                break;
        }
        return $output;
    }
    
    function upgrade_settings($old, $new) {
        foreach ($new as $key => $value) {
            if (!isset($old[$key])) $old[$key] = $value;
        }
        
        $unset = Array();
        foreach ($old as $key => $value) {
            if (!isset($new[$key])) $unset[] = $key;
        }
        
        foreach ($unset as $key) {
            unset($old[$key]);
        }
        
        return $old;
    }
    // helpers
    
	// initialization
    function lang_init() {
        $currentLocale = get_locale();
        if(!empty($currentLocale)) {
            $moFile = $this->plugin_path."/languages/".$this->plugin_name."-" . $currentLocale . ".mo";
            if (@file_exists($moFile) && is_readable($moFile)) load_textdomain($this->plugin_name, $moFile);
        }
    }
    
    function settings_menu() {
        add_submenu_page('options-general.php', $this->plugin_name, $this->plugin_name, $this->plugin_arguments["admin_page_settings"], $this->file_path, array(&$this, 'page_settings_panel'));
    }
    
    function plugin_menu() {
        add_submenu_page('plugins.php', $this->plugin_name, $this->plugin_name, $this->plugin_arguments["admin_page_plugin"], $this->file_path, array(&$this, 'page_plugins_panel'));
    }

    function manage_menu() {
        add_submenu_page('edit.php', $this->plugin_name, $this->plugin_name, $this->plugin_arguments["admin_page_manage"], $this->file_path, array(&$this, 'page_plugins_manage'));
    }
    
    function admin_head() {
        foreach ($this->extra_files_admin as $file) {
            echo $this->create_file_tag($file);
        }
    }
    
    function admin_footer() {
    }

    function wp_content($content) {
        return $content;
    }

    function wp_head() {
        foreach ($this->extra_files_wp as $file) {
            echo $this->create_file_tag($file);
        }
    }

    function wp_footer() {
    }
    // initialization
    
    // events & log
    function add_event($msg, $event = "init") {
        $this->events[] = Array(
            "time" => current_time('mysql'),
            "event" => $event,
            "plugin" => $this->plugin_name,
            "message" => $msg
        );
    }
    
    function dump_object($msg, $object) {
        $obj = print_r($object, true);
        $f = fopen($this->log_file, 'a+');
        fwrite ($f, sprintf("[%s] : %s\r\n", current_time('mysql'), $msg));
        fwrite ($f, "$obj");
        fwrite ($f, "\r\n");
        fclose($f);
    }
    
    function dump_string($msg, $event = "init") {
        $f = fopen($this->log_file, 'a+');
        fwrite ($f, sprintf("[%s] : %s : %s\r\n", current_time('mysql'), $event, $msg));
        fclose($f);
    }

    function dump_events() {
        $f = fopen($this->log_file, 'a+');
        foreach ($events as $e)
            fwrite ($f, sprintf("[%s] : %s : [%s] : %s<br />", $e["time"], $e["event"], $e["plugin"], $e["message"]));
        fclose($f);
    }
    
    function print_events($events) {
        foreach ($events as $e)
            echo sprintf("[%s] : %s : [%s] : %s<br />", $e["time"], $e["event"], $e["plugin"], $e["message"]);
    }
    // events & log
    
    // panels
    function page_settings_panel() {
    }

    function page_plugins_panel() {
    }

    function page_manage_panel() {
    }
    // panels
    
    // widget
    function widget_init() {
        $this->default_options_widget['wcore'] = 'wcore';

        if (!$options = get_option($this->plugin_arguments["widget_options"]))
            $options = array();
            
        $name = $this->plugin_name.' Widget';
        $widget_ops = array('classname' => $this->plugin_arguments["widget_options"], 'description' => $name);
        $control_ops = array('width' => $this->plugin_arguments["widget_width"], 'height' => $this->plugin_arguments["widget_height"], 'id_base' => $this->plugin_arguments["widget_idbase"]);
        
        $registered = false;
        foreach (array_keys($options) as $o) {
            if (!isset($options[$o]['wcore'])) continue;
                
            $id = $this->plugin_arguments["widget_idbase"].'-'.$o;
            $registered = true;
            wp_register_sidebar_widget($id, $name, array(&$this, 'widget_display'), $widget_ops, array( 'number' => $o ) );
            wp_register_widget_control($id, $name, array(&$this, 'widget_control'), $control_ops, array( 'number' => $o ) );
        }
        if (!$registered) {
            wp_register_sidebar_widget($this->plugin_arguments["widget_idbase"].'-1', $name, array(&$this, 'widget_display'), $widget_ops, array( 'number' => -1 ) );
            wp_register_widget_control($this->plugin_arguments["widget_idbase"].'-1', $name, array(&$this, 'widget_control'), $control_ops, array( 'number' => -1 ) );
        }
    }
    
    function widget_control($widget_args = 1){
        global $wp_registered_widgets;
        static $updated = false;

        if ( is_numeric($widget_args) )
            $widget_args = array('number' => $widget_args);
        $widget_args = wp_parse_args($widget_args, array('number' => -1));
        extract($widget_args, EXTR_SKIP);
        $options_all = get_option($this->plugin_arguments["widget_options"]);
        if (!is_array($options_all))
            $options_all = array();

        if (!$updated && !empty($_POST['sidebar'])) {
            $sidebar = (string)$_POST['sidebar'];
            $this->dump_object("POST", $_POST);
            
            $sidebars_widgets = wp_get_sidebars_widgets();
            if (isset($sidebars_widgets[$sidebar]))
                $this_sidebar =& $sidebars_widgets[$sidebar];
            else
                $this_sidebar = array();
            
            foreach ($this_sidebar as $_widget_id) {
                if ($this->plugin_arguments["widget_options"] == $wp_registered_widgets[$_widget_id]['callback'] && isset($wp_registered_widgets[$_widget_id]['params'][0]['number'])) {
                    $widget_number = $wp_registered_widgets[$_widget_id]['params'][0]['number'];
                    if (!in_array($this->plugin_arguments["widget_idbase"].'-'.$widget_number, $_POST['widget-id']))
                        unset($options_all[$widget_number]);
                }
            }
            foreach ((array)$_POST[$this->plugin_arguments["widget_form"]] as $widget_number => $posted) {
                if (!isset($posted['wcore']) && isset($options_all[$widget_number])) continue;
                $options = $this->widget_parse_options($posted);
                $options_all[$widget_number] = $options;
            }
            update_option($this->plugin_arguments["widget_options"], $options_all);
            $updated = true;
        }

        if (-1 == $number) {
            $this->options_widget_number = '%i%';
            $this->options_widget = $this->default_options_widget;
        }
        else {
            $o = $this->upgrade_settings($options_all[$number], $this->default_options_widget);
            $options_all[$number] = $o;
            update_option($this->plugin_arguments["widgetoptions"], $options_all);

            $this->options_widget_number = $number;
            $this->options_widget = $o;
        }
    }

    function widget_parse_options($post_params) {
        $options = array();
        $options['wcore'] = $post_params['wcore'];
        return $options;
    }

    function widget_display($args, $widget_args = 1) {
        extract($args);
        if (is_numeric($widget_args)) $widget_args = array('number' => $widget_args);
        $widget_args = wp_parse_args($widget_args, array( 'number' => -1 ));
        extract($widget_args, EXTR_SKIP);
        $options_all = get_option($this->plugin_arguments["widget_options"]);
        if (!isset($options_all[$number])) return;
        $this->options_widget = $options_all[$number];
    }
    // widget

    // widget form elements
    function get_widget_element_name($name) {
    	$el = sprintf("%s[%s][%s]", $this->plugin_arguments["widget_form"], $this->options_widget_number, $name);
    	return $el;
    }

    function get_widget_element_id($name) {
    	$el = sprintf("%s-%s", $this->plugin_arguments["widget_form"], $name);
    	return $el;
    }

    function get_widget_render_hidden($name, $value) {
    	$el = sprintf('<input type="hidden" id="%s" name="%s" value="%s" />',
    			$this->get_widget_element_id($name),
    			$this->get_widget_element_name($name),
    			$value
    		);
    	return $el;
    }

    function widget_element_name($name) {
    	echo $this->get_widget_element_name($name);
    }

    function widget_element_id($name) {
    	echo $this->get_widget_element_id($name);
    }

    function widget_render_hidden($name, $value) {
    	echo $this->get_widget_render_hidden($name, $value);
    }
    // widget form elements
}